Analyzers dynamic

xristy opened the pull request:
This PR implements configuration driven use of analyzers as suggested in [Issue 245](https://github.com/rnewson/couchdb-lucene/issues/245). 

AnalyzerTests.java is updated with appropriate tests and CONFIGURING_ANALYZERS.md documents the extension.

The org.json artifact was updated to 20160810.

The version was updated to 2.2.0-SNAPSHOT.

eroux commented:
@rnewson would you have time to review this PR?

rnewson commented:
Later this week hopefully, can't right now. Sounds awesome though!

Sent from my iPhone

> On 17 Apr 2017, at 13:27, Elie Roux <notifications@github.com> wrote:
> 
> @rnewson would you have time to review this PR?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

eroux commented:
Ok thanks a lot!

rnewson commented:
One quick comment. There's a lot of white space changes that make this look bigger than it is. Can you please update the relevant commits to eliminate this please? Notably in Analyzers.java itself. 

You also upgrade the json library version. A separate commit in the pr for that please, including reason. 

More later when I can focus, and thanks again for getting this started. 

Sent from my iPhone

> On 17 Apr 2017, at 13:48, Elie Roux <notifications@github.com> wrote:
> 
> Ok thanks a lot!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

xristy commented:
I've cleaned up Analyzers.java to reflect just the functional changes. 

I've reverted the org.json to 20090211. I had updated to the most recent simply to have access the source and docs since I was not familiar with the lib and there is no source associated with 20090211.

rnewson commented on CONFIGURING_ANALYZERS.md:
````diff
@@ -0,0 +1,134 @@
+```
+
+All of the examples so far refer to analyzers that are found in the Lucene distribution that is included with couchdb-lucene. All that is needed to add custom analyzers is to put the analyzer class and any dependencies (packaged in a jar typically) on the classpath of couchdb-lucene. A simple way of doing this is to put a jar with the custom analyzers in the `./lib` directory of the couchdb-lucen ...
+
````
this 'for example' doesn't appear to match the text.

rnewson commented on CONFIGURING_ANALYZERS.md:
```diff
@@ -0,0 +1,134 @@
+
+There are many other analyzers included in Lucene and there are occasions where custom analyzers are needed. 
+There is now support for configuring additional analyzers without needing to further modify couchdb-lucene. 
+This extension is a loose adaptation of the [eXist-db/Lucene integration](https://github.com/eXist-db/exist/blob/develop/extensions/indexes/lucene/src/org/exist/indexing/lucene/AnalyzerConfig.java)
```
thanks for mentioning this, we need to be very sure this isn't a derivative work as that exist-db code is licensed under LGPL and this project is not.

xristy replied:
I'm not sure what the test for "derivative work" is. I was aware of the functionality from having worked with eXist-db and intent on developing similar functionality in the context of couchdb-lucene. The core of the functionality is the use of native java reflection which then needed to be fit into the couchdb-lucene configuration and instantiation framework already in place.

I think the current [PR #249](https://github.com/rnewson/couchdb-lucene/pull/249) is largely independent of the [eXist-db code](https://github.com/eXist-db/exist/blob/develop/extensions/indexes/lucene/src/org/exist/indexing/lucene/AnalyzerConfig.java) and overlaps to the extent that there are limited ways in which java reflection is used to fetch a class constructor and invoke it.

rnewson commented on README.md:
```diff
@@ -164,7 +165,6 @@ Lucene has numerous ways of converting free-form text into tokens, these classes
 - standard
 - thai
 - whitespace
-- ngram
```
not sure why ngram moved in the README, strong preference for putting it back

xristy replied:
gratuitous attempt at alphabetizing. reverted in PR #249.

rnewson commented on README.md:
```diff
@@ -347,11 +352,11 @@ In addition to normal text-based range searches (using the "field:[lower TO uppe
+|long|`field<long>:[0 TO 100]`|
+|float|`field<float>:[0.0 TO 100.0]`|
+|double|`field<double>:[0.0 TO 100.0]`|
+|date|`field<date>:[from TO to]` where from and to match any of these patterns: `"yyyy-MM-dd'T'HH:mm:ssZ"`, `"yyyy-MM-dd'T'HH:mm:ss"`, `"yyyy-MM-ddZ"`, `"yyyy-MM-dd"`, `"yyyy-MM-dd'T'HH:mm:ss.SSSZ"`, `"yyyy-MM-dd'T'HH:mm:ss.SSS"`. So, in order to search for articles published in July, you would issue a following query: `published_at<date>:["2010-07-01T00:00:00"+TO+"2010-07-31T23:59:59"]`|
```
revert this hunk, it has nothing to do with the feature enhancement (a separate PR to improve the README would be fine though)

xristy replied:
Done

rnewson commented on pom.xml:
```diff
@@ -23,7 +23,7 @@
     <description>Full-text indexing for CouchDB</description>
     <url>http://github.com/rnewson/couchdb-lucene/</url>
-    <version>2.1.0-SNAPSHOT</version>
+    <version>2.2.0-SNAPSHOT</version>
```
revert this hunk, the feature enhancement doesn't mean we'll release as 2.2.0 and skip over 2.1.0. :)

xristy replied:
Done

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/couchdb/View.java:
```diff
@@ -17,6 +17,8 @@
 
 import com.github.rnewson.couchdb.lucene.util.Analyzers;
+import com.github.rnewson.couchdb.lucene.util.Constants;
+
```
ws only change, remove.

xristy replied:
ws and indentation issues should all be handled now.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Constants.java:
```diff
@@ -30,4 +30,12 @@ private Constants() {
 
     public static final String DEFAULT_FIELD = "default";
 
+	public static final String DEFAULT_ANALYZER = "standard";
```
indentation does not match existing style

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -51,114 +60,199 @@
         public Analyzer newAnalyzer(final String args) {
             return new BrazilianAnalyzer();
         }
+		
```
whitespace only change, remove.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -51,114 +60,199 @@
         public Analyzer newAnalyzer(final String args) {
             return new SmartChineseAnalyzer();
         }
+		
```
same comment for this and all the ones below. this code didn't change, so it should not appear in the diff.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -51,114 +60,199 @@
+		}
+		
+		@Override
+		public Analyzer newAnalyzer(final JSONObject json) throws JSONException {
```
please match indentation style of existing code

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+		
+		if (str.startsWith("{")) {
+			try {
+				return getAnalyzer( new JSONObject(str) );
```
remove whitespace within parens

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+			return getAnalyzer(Constants.DEFAULT_ANALYZER);
+		} 
+		
+		str = str.trim();
```
why trim here?

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+			try {
+				return getAnalyzer( new JSONObject(str) );
+			
+			} catch (Exception ex) {
```
catch Exception is quite broad, what is really thrown here that we should catch?

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+					} else {
+						return Analyzers.valueOf(key.toUpperCase()).newAnalyzer(args);
+					}
+				} catch (Exception ex) { }
```
what throws here? Is it always safe to ignore?

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+				} catch (Exception ex) { }
+			}
+
+			logger.error("Lucene index: analyzer class name is not defined");
```
include the class name.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+		try {
+			clazz = Class.forName(className);
+		} catch (ClassNotFoundException e) {
+			logger.error(String.format("Lucene index: analyzer class %s not found. (%s)", className, e.getMessage()));
```
include the exception as the second parameter to any logger call so we get it logged properly. i.e, 

`logger.error(String.format("analyzer class %s not found.", className), e);`

rnewson replied:
the 'lucene index' prefix in these log messages doesn't seem particularly helpful.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+
+		// Is the class an Analyzer?
+		if (!Analyzer.class.isAssignableFrom(clazz)) {
+			logger.error(String.format("Lucene index: analyzer class has to be a subclass of %s", Analyzer.class.getName()));
```
as I look at all these cases and their handling, I think a better approach would be to throw an exception, with a useful message, that we catch and log at the point where we call getAnalyzer(obj) above.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+		List<KeyTypedValue> cParams;
+		try {
+			cParams = getAllConstructorParameters(params);
+		} catch (Exception pe) {
```
this is getting to be a habit ;)

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+			cParams = getAllConstructorParameters(params);
+		} catch (Exception pe) {
+			logger.error(String.format("Unable to get parameters for %s: %s", className, pe.getMessage()), pe);
+			cParams = new ArrayList<>();
```
I don't think this is a good fallback, if the params are unparseable, we should fail, not continue by ignoring the attempt to specify parameters. Not indexing is preferable to silently indexing incorrectly.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+		String className = json.optString(Constants.CLASS);
+		JSONArray params = json.optJSONArray(Constants.PARAMS);
+
+		Analyzer newAnalyzer = null;
```
this var is used only in the final stage of the method, I suggest removing it.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+		}
+
+		// Create new analyzer
+		newAnalyzer = createInstance(clazz, cParamClasses, cParamValues);
```
rather than a method-scoped variable that is unused until the very end, an if statement here seems sufficient. tbh, given all we're doing is logging, which I think would be better done at the caller to this method anyway, I would just change this final section to 

`return createInstance(clazz, cParamClasses, cParamValues);`

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+
+			return (Analyzer) cstr.newInstance(vcParamValues);
+
+		} catch (IllegalArgumentException | IllegalAccessException | InstantiationException | InvocationTargetException | SecurityException e) {
```
heh, wow, I did not know that syntax, nice.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+			final Constructor<?> cstr = clazz.getDeclaredConstructor(vcParamClasses);
+			cstr.setAccessible(true);
+
+			if (logger.isDebugEnabled()) {
```
I'm not sure the isDebugEnabled is justified to avoid the very simple string format call.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+
+		try {
+			final Constructor<?> cstr = clazz.getDeclaredConstructor(vcParamClasses);
+			cstr.setAccessible(true);
```
it's been a while since I delved into reflection, do we need to setAccessible(true) here?

xristy replied:
no. removed. In principle would permit calling a private constructor which isn't sensible really.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+	 * @throws org.exist.indexing.lucene.AnalyzerConfig.ParameterException
+	 */
+	private static List<KeyTypedValue> getAllConstructorParameters(JSONArray params) throws ParameterException, JSONException {
+		final List<KeyTypedValue> parameters = new ArrayList<>();
```
s/parameters/result

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+			if (value == null) {
+				throw new ParameterException("The 'value' field of a string param must exist and must contain a String value.");
+			}
+
```
no need for blank line here

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+			}
+
+			parameter = new KeyTypedValue(name, value, String.class);
+
```
no need for blank line here

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+
+			try {
+				// ToDo: check where to close reader to prevent resource leakage
+				Reader fileReader = new java.io.FileReader(value);
```
yeah this needs to be to-done before merging

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+				parameter = new KeyTypedValue(name, fileReader, Reader.class);
+
+			} catch (java.io.FileNotFoundException ex) {
+				logger.error(String.format("File '%s' could not be found.", value), ex);
```
if the file isn't found, we need to error rather than continue

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+			break;
+		}
+
+		return parameter;
```
hrm. I would remove this 'parameter'. Just return in each clause of the switch statement if you have something to return, and return null in the default case. Much clearer that way.

rnewson commented on src/main/java/com/github/rnewson/couchdb/lucene/util/Analyzers.java:
```diff
@@ -186,13 +280,313 @@ protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComp
+	 * CLass for containing the Triple : key (name), corresponding value and
+	 * class type of value.
+	 */
+	private static class KeyTypedValue {
```
should also be final

xristy replied:
Thank you for your very thorough consideration of the pull request. I've created a new branch in my fork and made 3 clean commits: 1) implementing the analyzer configuration; 2) adding tests; and 3) adding documentation. I tried to answer all of the comments in the new branch. I was thinking to close this PR and make a new one with the new "clean" branch. I ended up making far too many changes to be comfortable trying to do everything with fixup and so on. Let me know if this approach will work for you.

xristy commented:
Thank you for your very thorough consideration of the pull request. I've created a new branch in my fork and made 3 clean commits: 1) implementing the analyzer configuration; 2) adding tests; and 3) adding documentation. I tried to answer all of the comments in the new branch. I was thinking to close this PR and make a new one with the new "clean" branch. I ended up making far too many changes to be comfortable trying to do everything with fixup and so on. Let me know if this approach will work for you.

rnewson commented:
hi, sorry for the delay in responding, I only get the weekends to hack on this project. Should I be reviewing PR 249 instead of this one?

xristy commented:
Yes please review PR # 249. I thought it would be more effective for me to create a new branch and PR than to try and repair commit history and such, especially since made more changes while tidying up the code. 

I've replied to several of your comments on this PR in the various comments but the actual changes I made are reflected in PR # 249. 

I appreciate very much the time you're taking to review this work. I think the extension should be useful to others.

rnewson commented:
Thanks. Will review new pr later today. 

Sent from my iPhone

> On 23 Apr 2017, at 16:20, Chris Tomlinson <notifications@github.com> wrote:
> 
> Yes please review PR # 249. I thought it would be more effective for me to create a new branch and PR than to try and repair commit history and such, especially since made more changes while tidying up the code.
> 
> I've replied to several of your comments on this PR in the various comments but the actual changes I made are reflected in PR # 249.
> 
> I appreciate very much the time you're taking to review this work. I think the extension should be useful to others.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

xristy commented:
This PR is superseded by PR 249
